src/lightning_utilities/core/imports.py

```diff
@@ -84,67 +84,48 @@ def compare_version(package: str, op: Callable, version: str, use_base_version:
 
 
 class RequirementCache:
-    """Boolean-like class for check of requirement with extras and version specifiers.
+    """Boolean-like class to check for requirement and module availability.
+
+    Args:
+        requirement: The requirement to check, version specifiers are allowed.
+        module: The optional module to try to import if the requirement check fails.
 
     >>> RequirementCache("torch>=0.1")
     Requirement 'torch>=0.1' met
     >>> bool(RequirementCache("torch>=0.1"))
     True
     >>> bool(RequirementCache("torch>100.0"))
     False
-    """
-
-    def __init__(self, requirement: str) -> None:
-        self.requirement = requirement
-
-    def _check_requirement(self) -> None:
-        if not hasattr(self, "available"):
-            try:
-                pkg_resources.require(self.requirement)
-                self.available = True
-                self.message = f"Requirement {self.requirement!r} met"
-            except Exception as ex:
-                self.available = False
-                self.message = f"{ex.__class__.__name__}: {ex}. HINT: Try running `pip install -U {self.requirement!r}`"
-
-    def __bool__(self) -> bool:
-        """Format as bool."""
-        self._check_requirement()
-        return self.available
-
-    def __str__(self) -> str:
-        """Format as string."""
-        self._check_requirement()
-        return self.message
-
-    def __repr__(self) -> str:
-        """Format as string."""
-        return self.__str__()
-
-
-class ModuleAvailableCache:
-    """Boolean-like class for check of module availability.
-
-    >>> ModuleAvailableCache("torch")
-    Module 'torch' available
-    >>> bool(ModuleAvailableCache("torch"))
+    >>> RequirementCache("torch")
+    Requirement 'torch' met
+    >>> bool(RequirementCache("torch"))
     True
-    >>> bool(ModuleAvailableCache("unknown_package"))
+    >>> bool(RequirementCache("unknown_package"))
     False
     """
 
-    def __init__(self, module: str) -> None:
+    def __init__(self, requirement: str, module: Optional[str] = None) -> None:
+        self.requirement = requirement
         self.module = module
 
     def _check_requirement(self) -> None:
```

Comment:
cc @awaelchli @Borda 

The idea is that we first try the old `RequirementCache` logic, and if it fails and we can, the old `ModuleAvailableCache` logic.

Reply:
I don't understand the approach. In the case we saw in the linked issue, the module was importable, but the requirement had a dependency mismatch of some sort. So how does this PR address this specific case? Would RequirementCache(x) return True because it falls back to the ModuleAvailableCache?

Reply:
Exactly. If `pkg_resources.require(requirement)` fails, then we try `module_available(module)`